### PR TITLE
chore: refactor object_store crate

### PR DESCRIPTION
**Pull request by pauldix**
This one is a bit of a yak shave in advance of me adding new functionality to the object store that I'll need to make the WAL Buffer Segment storage work.

This pulls the different backing implementations into their own modules. They're about to get more complex so it felt like it was time to separate them out rather than building towards a single multi-thousand line lib.rs. The error type is only defined in lib and imported by the individual modules, which I think makes it easier to work with.

This might be easiest to view commit by commit. The first refactor to pull apart the big lib.rs file is 3f593678e1beba5c327cd4822765e523257d6386.

The second commit adds a new `list_with_delimiter` method to object store and implements it for S3: 47f51dfbfb8381a8a93aa0e01b0a632f74fd4385. In the description of that commit I talk about the API design choice so have a look there to see if it makes sense.

And the third adds support for `list_with_delimiter` to the in-memory store: 9f5a1ec045b822d1edf0eab90f6057d5ac530651.

The object store interface still needs to be updated to have some wrapper object around paths as in #534

### CODE REVIEW

File: object_store/src/aws.rs [hunk 1]
```diff
@@ -0,0 +1,477 @@
+//! This module contains the IOx implementation for using S3 as the object
+//! store.
+use crate::{
+    Error, NoDataFromS3, Result, UnableToDeleteDataFromS3, UnableToGetDataFromS3,
```
Comment (alamb) [inline object_store/src/aws.rs:4]: Nice! You figured out a way to import the snafu stuff from other modules! 👍 I tried to figure this out a few months ago and I was continually confused about the difference between `Error::NoDataFromS3` and `NoDataFromS3`

**Review by alamb — APPROVED** [review]
I didn't carefully review that the code was copy/pasted correctly, but the general structure looks great to me.  I think this is a good step forward. 

I also really like the shared error pattern.

File: object_store/src/path.rs [hunk 2]
```diff
@@ -1,32 +1,63 @@
...
 /// directories in the object store.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
 pub struct ObjectStorePath {
     path: PathSource,
 }
 
+// Invariants to maintain/document/test:
+//
+// - always ends in DELIMITER
+// - does not contain unencoded DELIMITER
```
Comment (pauldix) [inline object_store/src/path.rs:20]: The full path should have the delimiters unencoded if it's for display or use in object storage. The directory or file names should have `/` encoded.

File: object_store/src/path.rs [hunk 3]
```diff
@@ -1,32 +1,63 @@
...
+// - for file paths: does not escape root dir
+// - for object storage: looks like directories
+
 impl ObjectStorePath {
     /// Pushes a part onto the path. Ensures the part is percent encoded
-    pub fn push(&mut self, part: &str) {}
+    // TODO: This would be nicer if `part` were `&str`, but how do we communicate
```
Comment (pauldix) [inline object_store/src/path.rs:26]: I was thinking `push` would be used to push on each individual directory. So server and snapshot should build the path with multiple push commands. That code is going to get refactored a bit while I'm doing the lifecycle. So I'd expect this to take `&str`.

File: object_store/src/path.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+/// Universal interface for handling paths and locations for objects and
+/// directories in the object store.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ObjectStorePath {
+    path: PathSource,
+}
+
+// Invariants to maintain/document/test:
+//
+// - always ends in DELIMITER
```
Comment (pauldix) [inline object_store/src/path.rs:19]: This is only true if it's a directory. If it's the end object, it should have some sort of file extension like `.parquet`, `.json`, or `.segment`

**Review by alamb — COMMENTED** [review]
FWIW I think the high level idea of this PR is good but that the PR is getting pretty large. I suggest that we get the functionality we have here into good shape, merge it to main, and then continue with additional work as a follow on PR.

File: object_store/src/path.rs [hunk 5]
```diff
@@ -0,0 +1,216 @@
+//! This module contains code for abstracting object locations that work
+//! across different backing implementations and platforms.
+
+use itertools::Itertools;
+use percent_encoding::{percent_encode, AsciiSet, CONTROLS};
+use snafu::Snafu;
+use std::path::PathBuf;
+
+/// A specialized `Result` for `ObjectStorePath`-related errors
+pub type Result<T, E = Error> = std::result::Result<T, E>;
```
[the remark this answers is not in the document]
Comment (carols10cents) [reply object_store/src/path.rs:10]: Whoops yep, good call, I had it fallible for a bit but switched it. I'll take these out.

**Comment by carols10cents** [comment]
The changes I made since the last reviews were the minimum needed to get CI passing (there are some ignored tests now because of the unimplementeds), so I'm going to merge and keep on moving :)

**Pull request merged**